### DISCUSSION TITLE
Shiny3 build issue when using Shiny.Push.FirebaseMessaging for Maui .NET7 IOS Firebase app?

### DISCUSSION THREAD

**Discussion by exquirentibus** [shinyorg/shiny#1340] [Q&A]
Using VS 2022 version 17.7.5,  I'm using Shiny 3.0.1 to add Push Notifications to a .NET7 Maui IOS app using Firebase (using the Apple Push Notification server (APNs)) and Entitlements.plist and Not the native Development and Production SSL Certifixcate (Apple Push Notification service SSL Certificates).

I can build and deploy to my iPhone when I only include 

<PackageReference Include="Shiny.Push" version="3.0.1" />
<PackageReference Include="Shiny.Hosting.Maui" version="3.0.1" />

But when I then include 
<PackageReference Include="Shiny.Push.FirebaseMessaging"  version="3.0.1" />

so that I can change my code in MauiProgram.cs from

builder.Services.AddPush<MauiApp.Delagates.MyPushDelegate>();

to: 

builder.Services.AddPushFirebaseMessaging<MauiApp.Delagates.MyPushDelegate>();

I get this build error:

C:\Program Files\dotnet\packs\Microsoft.iOS.Sdk\16.4.7107\tools\msbuild\iOS\Xamarin.Shared.targets(137,3): error : C:/Users/MyBox/AppData/Local/XamarinBuildDownloadCache/GAppM-8.9.1/GoogleAppMeasurement-8.9.1/Frameworks/GoogleAppMeasurement.xcframework has an incorrect or unknown format and cannot be processed.

I have my NuGets in C:\N to avoid the 240 char issue and have my Maui app in G:\Dev\xMauiApp\ so I don't think I'm hitting any limits with Google long names?

Possibly I'm off base to begin with and don't need to use Shiny.Push.FirebaseMessaging but I was trying to follow the VS Shiny Templates as best I could and I'm not seeing a break-point hit in OnTokenRefreshed (string token) in MyPushDelegate when I deploy to my iPhone and run in debug mode?  I'm assuming that where I'd get my token when the app runs?

What obvious stupid mistake am I making??

Thanks!!

**Comment by aritchie** [comment]
> C:\Program Files\dotnet\packs\Microsoft.iOS.Sdk\16.4.7107\tools\msbuild\iOS\Xamarin.Shared.targets(137,3): error : C:/Users/MyBox/AppData/Local/XamarinBuildDownloadCache/GAppM-8.9.1/GoogleAppMeasurement-8.9.1/Frameworks/GoogleAppMeasurement.xcframework has an incorrect or unknown format and cannot be processed.

Anything with a native binding seems to be suffering from this on windows builds right now.  Not sure what they are doing about it.  This problem does not exist if you build on mac.

> I'm not seeing a break-point hit in OnTokenRefreshed (string token) in MyPushDelegate

This is called when the token changes after use of RequestAccess.  Once push is granted, Shiny will attempt to auto-register across app starts  and if a token changes during that call, OnTokenRefresh is called.

**Reply by exquirentibus** [reply]
Will .NET8 with the Windows preview - work better for 'native binding'? 

For Firebase do I even need to use  Shiny.Push.FirebaseMessaging (which I can't get to build right now) so I won't have to deal with that?

And if I do Not need Shiny.Push.FirebaseMessaging - I'm unclear on how I get my token originally?  I really tried to follow your Firebase Maui template - but I'm obviously clueless as to the very basics?

BTW - seems like an Amazing set of libraries .. once I fully understand it!!!

**Comment by aritchie** [comment]
> Will .NET8 with the Windows preview - work better for 'native binding'?

Not sure.  I stay to mac right now for dev stuff

> For Firebase do I even need to use Shiny.Push.FirebaseMessaging (which I can't get to build right now) so I won't have to deal with that?

Not sure what you're asking me.  If you want to use Shiny with firebase on both iOS & Android - you need it.  If you want to just use native mechanics (do it yourself on the server), you can just use Shiny.Push.

> And if I do Not need Shiny.Push.FirebaseMessaging - I'm unclear on how I get my token originally? I really tried to follow your Firebase Maui template - but I'm obviously clueless as to the very basics?

There are samples that show how things work.  Specifically: https://github.com/shinyorg/shiny/blob/9ddaecda4ab7941e1acfc94ab7977e21deb10773/samples/Sample.Push.Maui/SetupViewModel.cs#L16

Basically, everything in Shiny begins with IManager.RequestAccess.

**Reply by exquirentibus** [reply]
I'm sorry to be so dense!  I'm just trying to Receive pushes on IOS from Firebase. 

My assumption is that I need to get a token and pass that to the server - who will be pushing out the notifications (but pushes could be from the Firebase Console initially).

You seem to be saying that the token is obtained during from the pushManager which is passed to the SetupViewModel constructor with dependency injection?

so var result = await this.pushManager.RequestAccess(); returns the token in 'result'?

If I'm correct so far .. do I need Shiny.Push.FirebaseMessaging to Receive messages from Firebase on IOS?

**Reply by aritchie** [reply]
Ok this is going beyond the spirit of the forum.  I would really recommend taking a look through the code to understand what it's doing and understand why you would want to use it.  The push module is quite small and should be easy to understand.  

Your assumptions are all correct.  Best of luck

**Reply by exquirentibus** [reply]
I just wanted to report that after updating to your 3.1.1 libraries with Windows VS 2022 17.8.0 Preview 6 .net7 I'm no longer getting that build error in Maui and can resume trying to make use of your wonderful library!!  

Shiny.Push.FirebaseMessaging, Shiny.Push and Shiny.Hosting.Maui are deploying to my iPhone on 17.1.  My Mac is running Ventura 13.6.1 with Xcode 15.0.1 and I only need to figure out how to properly use your libraries.

More steps before testing but .. Thanks so much for all your great work and for pointing out the most important areas of interest in your sample code!!!
